LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClientConnectionManager.cs

```diff
@@ -1,162 +1,399 @@
+
+                return connected;
             }
-        }
 
-        public void Release(LoRaDevice loRaDevice)
-        {
-            _ = loRaDevice ?? throw new ArgumentNullException(nameof(loRaDevice));
+            public Task DisconnectAsync() =>
+                DisconnectAsync(deferred: false);
+
+            private enum DisconnectionResult { Disconnected, Deferred, NotDisconnected }
 
-            if (this.managedConnections.TryRemove(GetConnectionCacheKey(loRaDevice.DevEUI), out var removedItem))
+            private Task<DisconnectionResult> DisconnectAsync(bool deferred) =>
+                InvokeExclusivelyAsync(doesNotRequireOpenConnection: true, async client =>
+                {
+                    DisconnectionResult result;
+
+                    if (Interlocked.Add(ref this.activities, 0) > 0)
+                    {
+                        this.shouldDisconnect = true;
+                        result = DisconnectionResult.Deferred;
+                    }
+                    else if (deferred && !this.shouldDisconnect)
+                    {
+                        result = DisconnectionResult.NotDisconnected;
+                    }
+                    else
+                    {
+                        this.shouldDisconnect = false;
+                        await client.DisconnectAsync();
+                        result = DisconnectionResult.Disconnected;
+                    }
+
+                    if (this.logger is { } logger && logger.IsEnabled(LogLevel.Debug))
+                    {
+                        const string message = $"{nameof(DisconnectAsync)} = {{DisconnectResult}}";
+                        logger.LogDebug(message, result);
+                    }
+
+                    return result;
+                });
+
+            public IAsyncDisposable BeginDeviceClientConnectionActivity()
             {
-                removedItem.Dispose();
+                var activities = Interlocked.Increment(ref this.activities);
+
+                if (this.logger is { } logger && logger.IsEnabled(LogLevel.Debug))
+                {
+                    const string message = $"{nameof(BeginDeviceClientConnectionActivity)} = {{Activities}}";
+                    logger.LogDebug(message, activities);
+                }
+
+                return new AsyncDisposable(async cancellationToken =>
+                {
+                    var activities = Interlocked.Decrement(ref this.activities);
+
+                    if (this.logger is { } logger && logger.IsEnabled(LogLevel.Debug))
+                    {
+                        const string message = $"{nameof(BeginDeviceClientConnectionActivity)} (disposal) = {{Activities}}";
+                        logger.LogDebug(message, activities);
+                    }
+
+                    if (activities > 0)
+                        return;
+
+                    _ = await DisconnectAsync(deferred: true);
```

Comment:
why would we disconnect if we do not have any more activities? The owning LNS should not drop the connection.

Reply:
This is only dealing with _deferred_ disconnects (`deferred: true`). It means, someone issued a `DisconnectAsync` while activities were taking place. To avoid interference, the actual disconnection is deferred until all activities are done. I agree that this can mislead to read that a disconnection is done, but it's predicated. I could rename the parameter to `ifDeferred` or `whenDeferred` if it helps?

Reply:
> The owning LNS should not drop the connection.

The ownership problem is not being solved directly in this PR. It's preparing the terrain to make that easier to fit in a more deterministic setup.

Reply:
After talking to Atif I understand: This was added to support a longer operation with a disconnect in between, like the join case (disconnecting in the factory and then defer that operation until we are done saving the twin). 

Smart! <3